crates/dbs-legacy-devices/src/rtc_pl031.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 Alibaba Cloud. All Rights Reserved.
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! ARM PL031 Real Time Clock
+//!
+//! This module implements a PL031 Real Time Clock (RTC) that provides to provides long time base counter.
+use std::convert::TryInto;
+
+use dbs_device::{DeviceIoMut, IoAddress};
+use dbs_utils::metric::{IncMetric, SharedIncMetric};
+use log::warn;
+use vm_superio::rtc_pl031::{Rtc, RtcEvents};
+
+/// Metrics specific to the RTC device
+#[derive(Default)]
+pub struct RTCDeviceMetrics {
+    /// Errors triggered while using the RTC device.
+    pub error_count: SharedIncMetric,
+    /// Number of superfluous read intents on this RTC device.
+    pub missed_read_count: SharedIncMetric,
+    /// Number of superfluous write intents on this RTC device.
+    pub missed_write_count: SharedIncMetric,
+}
+
+impl RtcEvents for RTCDeviceMetrics {
+    fn invalid_read(&self) {
+        self.missed_read_count.inc();
+        self.error_count.inc();
+    }
+
+    fn invalid_write(&self) {
+        self.missed_write_count.inc();
+        self.error_count.inc();
+    }
+}
+
+/// The wrapper of Rtc struct to implement DeviceIoMut trait.
+pub struct RTCDevice {
+    pub rtc: Rtc<RTCDeviceMetrics>,
+}
+
+impl RTCDevice {
+    pub fn new() -> Self {
+        let metrics = RTCDeviceMetrics::default();
+        Self {
+            rtc: Rtc::with_events(metrics),
+        }
+    }
+}
+
+impl DeviceIoMut for RTCDevice {
+    fn read(&mut self, _base: IoAddress, offset: IoAddress, data: &mut [u8]) {
+        if data.len() == 4 {
+            self.rtc
+                .read(offset.raw_value() as u16, data.try_into().unwrap())
+        } else {
+            warn!(
+                "Invalid RTC PL031 read: offset {}, data length {}",
+                offset.raw_value(),
+                data.len()
+            );
+            self.rtc.events().invalid_read();
+        }
+    }
+
+    fn write(&mut self, _base: IoAddress, offset: IoAddress, data: &[u8]) {
```

Comment:
Why `read` use `&mut [u8]` (which I don't see the usage in `read` function)  and `write` use `&[u8]`?

Reply:
Got that